src/server.ts

```diff
@@ -0,0 +1,195 @@
+    }
+
+    const Component = new ComponentClass(ownProps, context);
+
+    let newContext = context;
+    if (Component.getChildContext) newContext = assign({}, context, Component.getChildContext());
+
+    if (!store && ownProps.store) store = ownProps.store;
+    if (!store && newContext.store) store = newContext.store;
+
+    if (!client && ownProps.client && ownProps.client instanceof ApolloClient) {
+      client = ownProps.client as ApolloClient;
+    }
+    if (!client && newContext.client && newContext.client instanceof ApolloClient) {
+      client = newContext.client as ApolloClient;
+    }
+
+    getQueriesFromTree({
+      component: getChildFromComponent(Component),
+      context: newContext,
+      queries,
+    });
+  } else if (props && props.children) {
+    Children.forEach(props.children, (child: any) => getQueriesFromTree({
+      component: child,
+      context,
+      queries,
+    }));
+  }
+
+  return { queries, client, store };
+}
+
+// XXX component Cache
+export function getDataFromTree(app, ctx: any = {}): Promise<any> {
+
+  let { client, store, queries } = getQueriesFromTree({ component: app, context: ctx });
+
+  if (!store && client && !client.store) client.initStore();
+  if (!store && client && client.store) store = client.store;
+  // no client found, nothing to do
+  if (!client || !store) return Promise.resolve(null);
+
+  // no queries found, nothing to do
+  if (!queries.length) return Promise.resolve({ store, client, initialState: store.getState() });
+
+  // run through all queries we can
+  return processQueries(queries, client)
+      .then(trees => Promise.all(trees.map(x => {
+          const { component, ownProps, context } = x;
+          if (!component) return;
+          // Traverse wrapped components of resulting queries
+          // NOTE: sub component queries may fire again,
+          // but they will just return back existing data
+          const Element = createElement(component, ownProps) as any;
+          const child = getChildFromComponent(Element && new Element.type(ownProps, context));
+          if (!child) return;
+
+          // traverse children nodes
+          return getDataFromTree(child, context);
+      })))
+      .then(() => ({ store, client, initialState: store.getState() }));
+
+}
+
+export function renderToStringWithData(component) {
+  return getDataFromTree(component)
+    .then(({ store, client }) => {
+      let markup = ReactDOM.renderToString(component);
+      let initialState = store.getState();
+      const key = client.reduxRootKey;
+      // XXX apollo client requires a lot in the store
+      // can we make this samller?
+      for (let queryId in initialState[key].queries) {
+        let fieldsToNotShip = ['minimizedQuery', 'minimizedQueryString'];
+        for (let field of fieldsToNotShip)  delete initialState[key].queries[queryId][field];
+      }
+      initialState = encodeURI(JSON.stringify(initialState));
```

Comment:
I assume encodeURI is wrong? It'll result in
`Uncaught SyntaxError: Unexpected token %`